### PR TITLE
Remove duplication of npm install and npm dedupe in CircleCI

### PR DESCRIPTION
**Pull request by josemigallas**
`npm install` is run twice during each workflow, in `dependencies_npm` and `assets_precompile` jobs.

It takes a lot of time to complete even if modules are present. However, NPM modules can be safely shared between jobs and workflows until package-lock.json changes.

This way we can reduce the times it is run to the minimum, by using CircleCI cache properly.

Also `npm dedupe` is run twice, one during `dependencies_lint` and then again in `lint`. Simplified structure will be already stored in the cache therefore license_check should not complain about duplicated dependencies.

### CODE REVIEW

**Comment by mikz** [comment]
There is no `npm check` ? Because then you can use that like `npm check || npm install`.

https://circleci.com/docs/2.0/caching/ describes how you can take a cache from the same branch and just do partial install, or even from master. So you are not installing all dependencies, but just some.

File: openshift/system/Dockerfile [hunk 1]
```diff
@@ -43,7 +43,7 @@ ENV BASH_ENV=/opt/system/etc/scl_enable \
 RUN export ${BUNDLER_ENV} >/dev/null \
     && source /opt/system/etc/scl_enable \
     && gem install bundler --version 1.16.2 \
-    && bundle install --deployment --jobs $(grep -c processor /proc/cpuinfo) --retry=5
+    && bundle install --deployment --jobs $(grep -c processor /proc/cpuinfo) --retry=5
```
Comment (mikz) [inline openshift/system/Dockerfile]: `git add -p` would show you this unnecessary change

File: .circleci/config.yml [hunk 2]
```diff
@@ -338,6 +338,7 @@ jobs:
           name: Install NPM dependencies
           command: |
             npm --version
+            npm set progress false
```
Comment (mikz) [inline .circleci/config.yml]: can this be controlled by ENV ?

Comment (josemigallas) [reply .circleci/config.yml]: Well... I think this line might be actually redundant (https://docs.npmjs.com/misc/config#progress).

**Comment by josemigallas** [comment]
@mikz Thanks for reviewing.

I did explain myself poorly, I've updated the description already. NPM does not download again modules that are already present but it takes a lot of time auditing them. Running this once and storing it in the cache as you pointed out would be enough and much faster.

File: openshift/system/Dockerfile [hunk 3]
```diff
@@ -57,6 +57,8 @@ RUN export ${BUNDLER_ENV} >/dev/null \
     && chmod g+w -vfR log tmp public/assets db/sphinx \
     && umask 0002 \ 
     && cd /opt/system \
+    && npm set progress false \
+    && npm install \
```
Comment (mikz) [inline openshift/system/Dockerfile]: This can be done sooner. Just just bundler just adds Gemfile and runs this can be just add package-lock.json and run npm and be own layer.

Comment (josemigallas) [reply openshift/system/Dockerfile]: So, I tried to move this up but then `npm` is not found. See [#38939](https://circleci.com/gh/3scale/porta/38939) and [#38871](https://circleci.com/gh/3scale/porta/38871) for instance.

File: openshift/system/Dockerfile.on_prem [hunk 4]
```diff
@@ -58,6 +59,8 @@ RUN export ${BUNDLER_ENV} >/dev/null \
     && chmod g+w -vfR log tmp public/assets db/sphinx \
     && umask 0002 \
     && cd /opt/system \
+    && npm set progress false \
+    && npm install \
```
Comment (didierofrivia) [inline openshift/system/Dockerfile.on_prem]: ` npm install --no-progress` (?)

Comment (didierofrivia) [reply openshift/system/Dockerfile.on_prem]: Saw the other comment... does is it set CI env then?

Comment (josemigallas) [reply openshift/system/Dockerfile.on_prem]: It must be set in our circleci build but this commands are executed inside the docker container, don't they?
In any case it looks better with `--no-progress`.
